### PR TITLE
duplicates calls to remove eds cluster

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/19098

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@howardjohn  I think this should take care of duplicate calls?

**Comment by ramaraochavali** [comment]
@hzxuzhonghu Can you PTAL - Does this fix sound correct to you?

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -685,10 +685,6 @@ func (s *DiscoveryServer) removeCon(conID string, con *XdsConnection) {
 	adsClientsMutex.Lock()
 	defer adsClientsMutex.Unlock()
 
-	for _, c := range con.Clusters {
-		s.removeEdsCon(c, conID)
```
Comment (hzxuzhonghu) [inline pilot/pkg/proxy/envoy/v2/ads.go]: You can not do this, since this may leads to memory leak

**Comment by hzxuzhonghu** [comment]
@ramaraochavali Actually I can not figure out how this happened.

**Comment by hzxuzhonghu** [comment]
Oh, maybe there is a data race on `EdsClients`

```
		c.mutex.Lock()
		delete(c.EdsClients, connection.ConID)
		c.mutex.Unlock()
		if len(c.EdsClients) == 0 {
			// This happens when a previously used cluster is no longer watched by any
			// sidecar. It should not happen very often - normally all clusters are sent
			// in CDS requests to all sidecars. It may happen if all connections are closed.
			adsLog.Debugf("EDS: Remove unwatched cluster node:%s cluster:%s", connection.ConID, rc)
			delete(edsClusters, rc)
		}
```

**Comment by ramaraochavali** [comment]
@hzxuzhonghu Yeah, I think the the lock should not be cleared before we delete - I think that was the mistake. Does the new change sound right to you?

**Comment by hzxuzhonghu** [comment]
Look at this, EdsClients can at most include one key-value https://github.com/istio/istio/blob/7a48624126bb9283a0182ae53e24d9cacf3d86fb/pilot/pkg/proxy/envoy/v2/eds.go#L866
This is the root cause.

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 2]
```diff
@@ -842,8 +842,8 @@ func (s *DiscoveryServer) updateEdsClients(added sets.Set, removed sets.Set, con
 			continue
 		}
 		c.mutex.Lock()
+		defer c.mutex.Unlock()
```
Comment (hzxuzhonghu) [inline pilot/pkg/proxy/envoy/v2/eds.go]: This is not the cause, as L835 edsClusterMutex.Lock() can protect it. But i suggest fix it here too.

c.mutex.Lock()
delete(c.EdsClients, connection.ConID)
clients := len(c.EdsClients)
c.mutex.Unlock()

if clients == 0 {
...

Comment (ramaraochavali) [reply pilot/pkg/proxy/envoy/v2/eds.go]: @hzxuzhonghu ok. Thanks. Fixed it. PTAL

**Comment by hzxuzhonghu** [comment]
can you fix this https://github.com/istio/istio/pull/19100#issuecomment-557347981

**Comment by ramaraochavali** [comment]
@hzxuzhonghu I missed your other comment, sorry, but I do not fully understand it though - But I found one issue, fixed it - Is that your comment too?

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 3]
```diff
@@ -860,12 +861,13 @@ func (s *DiscoveryServer) updateEdsClients(added sets.Set, removed sets.Set, con
...
-			// Currently use edsClusterMutex lock
-			c.mutex.Lock()
-			c.EdsClients[connection.ConID] = connection
-			c.mutex.Unlock()
 		}
+
+		// TODO: find a more efficient way to make edsClusters and EdsClients init atomic
+		// Currently use edsClusterMutex lock
+		c.mutex.Lock()
+		c.EdsClients[connection.ConID] = connection
```
Comment (hzxuzhonghu) [inline pilot/pkg/proxy/envoy/v2/eds.go]: yes, this is what i mean, the previous pr introduced a bug.

Comment (ramaraochavali) [reply pilot/pkg/proxy/envoy/v2/eds.go]: ok. cool

**Pull request merged**